substrate/primitives/runtime-interface/src/polkavm.rs

```diff
@@ -0,0 +1,30 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub use polkavm_derive::{polkavm_export, polkavm_import};
+
+#[polkavm_derive::polkavm_define_abi(allow_extra_input_registers)]
+pub mod polkavm_abi {}
```

Comment:
What does that do? :D

Reply:
It's a magic macro. (;

Basically, it defines the trait machinery required to automatically convert types on an FFI boundary. It essentially generates this:

```rust
pub mod polkavm_abi {
    pub trait IntoHost {
        type Regs;
        type Destructor;

        fn into_host(value: Self) -> (Self::Regs, Self::Destructor);
    }

    pub trait FromHost {
        type Regs;
        fn from_host(value: Self::Regs) -> Self;
    }

    #[doc(hidden)]
    pub mod private {
        /* ...more stuff... */
    }
}
```

The idea here is that you can define how the types are marshaled through the FFI boundary in one place, and then when you use the `polkavm_import` or `polkavm_export` macro you can give it a path to this module, and it will use the `IntoHost` and `FromHost` traits defined within to do the conversions. For example, something like this:

```rust
#[polkavm_derive::polkavm_define_abi]
pub mod my_abi {}

impl self::my_abi::FromHost for FancyType {
    type Regs = (u32, u32,); // It needs two registers to be marshaled through the FFI boundary.
    fn from_host(value: Self::Regs) -> Self { ... }
}

#[polkavm_derive::polkavm_import(abi = my_abi)]
extern {
    fn hostcall() -> FancyType;
}

fn get_fancy() {
    let fancy = unsafe { hostcall(); }
}
```

This is somewhat similar to the traits we already have in Substrate for converting into/from WASM FFI types, except it's part of the upstream instead of being Substrate-specific, and you can have multiple ABIs defined (e.g. you could have one for Substrate runtimes and another for smart contracts, and they won't conflict with each other).